src/group_theory/iwasawa.lean

```diff
@@ -0,0 +1,487 @@
+
+end subgroup
+end Maximal_Subgroups
+
+section Basic_Lemmas
+/-- If G and H have multiplications *
+and φ : G → H is a surjective multiplicative map,
+and if G is commutative, then H is commutative.
+
+Since I only use it with groups,
+I should probably use function.surjective.comm_semigroup
+--/
+lemma surj_to_comm {G H : Type*} [has_mul G] [has_mul H] (φ: mul_hom G H) :
+   function.surjective φ → is_commutative G (*) → is_commutative H (*) :=
+begin
+  intros is_surj is_comm,
+  apply is_commutative.mk,
+  intros a b,
+    obtain ⟨a', ha'⟩ := is_surj a, obtain ⟨b', hb'⟩ := is_surj b,
+    rw ← ha', rw ← hb',
+    let z := ⇑φ, let z₂ := φ.to_fun, have : z = z₂  , by refl,
+    rw ← mul_hom.map_mul φ a' b',
+    rw ← mul_hom.map_mul φ b' a',
+    apply φ.congr_arg,
+    refine is_commutative.cases_on is_comm _, intro,
+    exact comm a' b',
+end
+
+/- No more used
+
+/-- f(f⁻¹(t)) = f.range ∩ t  -/
+lemma image_preimage_eq {α β : Type*} (f: α → β) (t: set β):
+  set.image f (set.preimage f t) = set.range f ∩ t :=
+begin
+    sorry,
+end
+
+lemma image_preimage_eq' {α β : Type*} (f: α → β) (t: set β):
+  set.image f (set.preimage f t) = set.range f ∩ t :=
+  by rw [← set.image_univ, ← set.image_inter_preimage f set.univ t, set.univ_inter]
+
+/-- If f is surjective, then f(f⁻¹(t)) = t -/
+--  is in mathlib :  set.image_preimage_eq
+lemma image_preimage_eq_of_surjective' {α β : Type*} (f : α → β) (t : set β) :
+  function.surjective f → set.image f (set.preimage f t) = t :=  λ hf,
+  by rw [image_preimage_eq f t, set.range_iff_surjective.2 hf,  set.univ_inter]
+ -/
+end Basic_Lemmas
+
+section Primitive_And_Transitive_Actions
+open_locale big_operators pointwise
+open function
+
+open set mul_action subgroup
+
+variables (G : Type*) [group G] (X : Type*) [mul_action G X]
+
+/-- Transitive actions -/
+structure is_transitive
+extends is_pretransitive G X : Prop :=
+(is_nonempty : nonempty X)
+
+/-
+lemma nonempty_of_transitive : is_transitive G X → nonempty X
+:= λ h, h.is_nonempty
+
+lemma pretransitive_of_transitive : is_transitive G X → is_pretransitive G X :=
+λ h, h.is_pretransitive
+-/
+
+variables (G X)
+structure is_primitive
+extends is_transitive G X : Prop :=
+(has_maximal_stabilizers: ∀ x:X, (stabilizer G x).is_maximal)
+
+/-
+lemma nonempty_of_primitive : is_transitive G X → nonempty X
+:= λ h, h.is_nonempty
+
+lemma transitive_of_primitive : is_primitive G X → is_transitive G X
+:= λ h, is_transitive.mk h.is_nonempty h.is_pretransitive
+
+lemma has_maximal_stabilizer : is_primitive G X →  ∀ (x : X), (stabilizer G x).is_maximal
+:= λ h, h.has_maximal_stabilizers
+-/
+
+/-- A 2-transitive action is primitive -/
+/- Part of the proof establishes that stabilizers of n-transitive
+actions are (n-1)-transitive. Todo: rewrite using this. -/
+theorem is_primitive_of_two_trans
+  : (∀ (x₁ x₂ y₁ y₂ : X) (hx: x₁ ≠ x₂) (hy: y₁ ≠ y₂), ∃ g: G,
+  g • x₁  = y₁ ∧ g • x₂ = y₂ ) →
+  nontrivial X → is_primitive G X :=
```

Comment:
I haven't tried this on my computer, but I suspect this might work. It marks `nontrivial X` as an instance, so that Lean knows to look for it.
```suggestion
theorem is_primitive_of_two_trans (h2X: ∀ (x₁ x₂ y₁ y₂ : X) (hx: x₁ ≠ x₂) (hy: y₁ ≠ y₂),
  ∃ g: G, g • x₁ = y₁ ∧ g • x₂ = y₂) [hX : nontrivial X] : is_primitive G X :=
```

Reply:
That works, thanks !

Reply:
Great. Can you push the change then?